src/calliope/model.py

```diff
@@ -448,22 +448,9 @@ def solve(self, force: bool = False, warmstart: bool = False, **kwargs) -> None:
 
         solver_config = update_then_validate_config("solve", self.config, **kwargs)
 
-        shadow_prices = solver_config["shadow_prices"]
-
-        if shadow_prices:
-            for constraint_name in shadow_prices:
-                if constraint_name not in self.math.constraints.keys():
-                    shadow_prices.remove(constraint_name)
-                    exceptions.warn(
-                        f"{constraint_name} was listed in `config.solve.shadow_prices`"
-                        "but is not a valid constraint. Shadow prices for this"
-                        "constraint will not be tracked.",
-                        exceptions.ModelWarning,
-                    )
-            # Only actually activate shadow price tracking if at least one valid
-            # constraint remains in the list after filtering out invalid ones
-            if shadow_prices:
-                self.backend.shadow_prices.activate()
+        updated_shadow_prices = self.backend._activate_shadow_prices_if_needed(
```

Comment:
I wouldn't return this information to the model object. Instead, keep it within the backend object. In `backend.BackendModel.load_results` you can attach the relevant shadow price arrays at the same time as adding the resolved variable and global expression arrays.